File: gridTraveler.rb
# Grid Traveller using RECURSION

def grid_traveler(x,y)
  return 1 if x == 1 && y == 1
  return 0 if x == 0 || y == 0
  grid_traveler(x - 1,y) + grid_traveler(x,y - 1)
end


# Grid Traveller using MEMOIZATION && RECURSION
def grid_traveler(x, y, memo={})
  key = x,y
  return memo[key] if memo.include?(key)
  return 1 if x == 1 && y == 1
  return 0 if x == 0 || y == 0

  memo[key] = grid_traveler(x - 1, y, memo) + grid_traveler(x, y - 1, memo)

  return memo[key]
end

File: gridTraveler.js
const grid_traveler = (x,y) =>{
  const array = Array(x + 1)
  .fill().map(() => Array(y + 1).fill(0))

  array[1][1] = 1

  for (let i = 0; i <= x; i++){
    for (let j = 0; j <= y; j++){
      const current = array[i][j]
      if (i + 1 <= x) array[i + 1][j] += current
      if (i + 1 <= x) array[i][j + 1] += current
    }
  }
  return array[x][y]
}
 console.log(grid_traveler(3,4))
 console.log(grid_traveler(18,14))
 console.log(grid_traveler(13,4))
 console.log(grid_traveler(18,4))
